lib/validator/v3/validator.go

```diff
@@ -1342,6 +1366,64 @@ func validateRouteTableRange(structLevel validator.StructLevel) {
 	}
 }
 
+func validateBGPCommunities(structLevel validator.StructLevel) {
+	cl := structLevel.Current().Interface().(api.CommunityKVPair)
+
+	isValid := isValidCommunity(cl.Value, "Spec.Communities[].Value", structLevel)
+	if !isValid {
+		log.Warningf("community value is invalid: %v", cl.Value)
+		structLevel.ReportError(reflect.ValueOf(cl.Value), "Spec.Communities[].Value", "",
+			reason("invalid community value or format used."), "")
+	}
+}
+
+func validatePrefixAdvertisements(structLevel validator.StructLevel) {
+	pa := structLevel.Current().Interface().(api.PrefixAdvertisements)
+
+	_, _, err := cnet.ParseCIDROrIP(pa.CIDR)
+	if err != nil {
+		log.Warningf("CIDR value is invalid: %v", pa.CIDR)
+		structLevel.ReportError(reflect.ValueOf(pa.CIDR), "Spec.PrefixAdvertisements[].CIDR", "",
+			reason("invalid CIDR value."), "")
+	}
+
+	// Only validates PrefixAdvertisements[].Communities[] values that follow `aa:nn` or `aa:nn:mm` format,
+	// PrefixAdvertisements[].Communities[] that refers values set in Spec.Communities will be validated in clientv3.bgpconfig
+	for _, v := range pa.Communities {
+		isValidCommunity(v, "Spec.PrefixAdvertisements[].Communities[]", structLevel)
+	}
+}
+
+func isValidCommunity(communityValue string, fieldName string, structLevel validator.StructLevel) bool {
+	if standardCommunity.MatchString(communityValue) {
+		validateCommunityValue(communityValue, fieldName, structLevel, false)
+	} else if largeCommunity.MatchString(communityValue) {
+		validateCommunityValue(communityValue, fieldName, structLevel, true)
+	} else {
+		return false
```

Comment:
Do we need to `structLevel.ReportError` in this branch?

Reply:
Not for the else condition, `isValidCommunity` is called in 2 places, one while validating `specs.community` and other while validating `specs.prefixAdvertisements.community`, in the later case, if it is not a valid community it could be a community name that is already defined (we do this validation in lib/clientv3/bgpconfig), and calling function takes care of `structLevel.ReportError` in former case.